### DISCUSSION TITLE
.Cls file ATTRIBUTE options

### DISCUSSION THREAD

**Discussion by mburns08109** [twinbasic/twinbasic#432] [Ideas]
I recently re-stumbled across the ATTRIBUTE lines which can be found in .cls files (at least).
This is the same feature that permits NEWENUM() to work with FOR EACH...NEXT constructs and which requires VBA to export and re-import the .cls file to implement.

My question is two-fold.
1) is there an "official" comprehensive source for what actual attributes are supported for VBA(at least)? I found some pages online that seem to show at least a partial list, but nothing from MS with what looks to be "Official" documentation on this subject.

2) The reason I'm asking is that I was playing around with a new custom collection class in VBA, and stumbled into the VB_Description and VB_VarDescription attributes, and like what they do for making code more self-documenting (especially in the F2 object explorer). Wouldn't it be nice to have similar options for ALL the potential VBx/tB constructs like that? VB_EnumDescription, VB_TypeDescription, ...and options for the members of each of those... to name a few possibilities.

In short, if it could conceivably show up in an F2 Object explorer list, I'd like to be able to slap at least a description on it...and it'd be really great for some syntax support that didn't require an export/import cycle like VBA does to get it.

**Comment by DavidHollman** [comment]
Not official but a decent descriptive list of many of the attributes: https://stackoverflow.com/a/34688164/3195477

**Reply by mburns08109** [reply]
I just want to call out the linked sub-document which details the createability of a VBA class via the use of these ATTRIBUBTES.
https://docs.microsoft.com/en-us/openspecs/microsoft_general_purpose_programming_languages/ms-vbal/a09fd48e-abed-4da8-8c4c-a110bf4ef6b6?redirectedfrom=MSDN

**Reply by mburns08109** [reply]
There was also a little hint at the bottom of that linked article concerning possible different uses or settings for those attributes that may be applied to "Library" code/class modules.

I would LOVE to find some more details on that topic as well - and then we can think about how it may (or may not) affect tB code.
For example: could tB code be used as a library module/add-in in VBA for Access/Excel? (it would be a GREAT way to help secure IP for Add-In builders.)

**Comment by mburns08109** [comment]
There are some additional details to be found here:
https://www.vitoshacademy.com/vba-attributes-the-high-end-in-vba/

One of the key items in here is "Attribute <method/propertyName>.UserMemId = 0"  <-- sets <method/propertyName> as the class default method/property
Attribute <methodName>.VB_UserMemId = -4  <-- used to set the NewEnum() as the Enum iterator method flag

Even more details can be found here:
https://christopherjmcclellan.wordpress.com/2015/04/21/vb-attributes-what-are-they-and-why-should-we-use-them/

...but still I would not call this combination a "comprehensive" list.

**Comment by bclothier** [comment]
Just in case:

The VB's attribute would be considered legacy in tB language and you should use the new attribute keyword instead. See #25.

**Comment by WaynePhillipsEA** [comment]
Many of the VBx attributes are already supported in tB, but as @bclothier rightly mentions, these are considered legacy syntax as we now have much nicer attributes (#25).   e.g:

`Attribute <blah>.UserMemId = 0`      becomes   `[ DefaultMember ]`     (above the procedure definition)
`Attribute <blah>.UserMemId = -4`      becomes   `[ Enumerator ]`     (above the procedure definition)
`Attribute <blah>.UserMemId = 555`      becomes    `[ DispId(555) ]`     (above the procedure definition)
`Attribute <blah>.Description = "blah"`     becomes   `[ Description("blah") ]`     (above the procedure definition)

>  Wouldn't it be nice to have similar options for ALL the potential VBx/tB constructs like that? VB_EnumDescription, VB_TypeDescription

The new `[ Description("") ]` attribute is supported for all element types.   Although we don't have a project explorer yet, the description will be shown as you hover over references to those elements in your code, and is also shown in intellisense / signature-help scenarios.

**Reply by mburns08109** [reply]
You were aware of the class instancing stuff using attributes already? (because this was kind-of news to me, but the last I really was needing to delve into that stuff was back when I was heavy in VB6...an eon or so ago - and we have class properties for that in the VB6 IDE)

**Reply by WaynePhillipsEA** [reply]
Yes, they are already supported, and have equivalent tB syntax as follows;

`Attribute VB_Exposed = True` becomes `Public Class`      (default)
`Attribute VB_Exposed = False` becomes `Private Class`
`Attribute VB_Creatable = True` becomes `[ COMCreatable(True) ]`     (default)
`Attribute VB_Creatable = False` becomes `[ COMCreatable(False) ]`

**Reply by mburns08109** [reply]
Hmm...according to the article linked above...VBx (or is it just VBA?)'s default is Private, not Public.
Should that be a discussion point? (or was it already and I missed it?)

**Reply by WaynePhillipsEA** [reply]
The change won't affect legacy code, as tB will pick up the legacy attribute (VB_Exposed) from the CLS file.   It could cause some confusion if the developer moves the code over to a twin file (in a Class block), and I don't think this difference has yet been discussed.

The reason this was missed was because we only actually starting supporting Private classes recently in tB, and so the default had to be Public in order to allow creation of ActiveX / COM libraries in tB in the earlier builds where we didn't allow Private/Public to be specified.   That said, I'm willing to consider a potential breaking change here.

**Reply by mburns08109** [reply]
OK, I just looked again, and yeah - that document is specific to VBA. So, I guess it makes sense that VBA defaults to Private. I've only really see the effect of this setting in Excel VBA - where the project explorer can list multiple different VBA projects at once - one for each open workbook. MS clearly didn't care to expose this functionality through the VBA editor or VBA language intended for the casual developer.

I need to think on the reasons for that - and the impact of it a bit more.

I can see it affecting Excel VBA code though - and also affecting (possibly positively - except from a code security POV perhaps) Add-In/"Library" projects.

**Reply by mburns08109** [reply]
Perhaps the key question here is: what do we envision tB's impact on using VBA to be?  Will there be scenarios where tB can (practically, mind you) used in place of/instead of VBA? ...or is the current thinking more along the lines of tB being used "alongside" of VBA (as in - as "Library" project code)?

**Reply by mburns08109** [reply]
Perhaps one answer may be a project-level setting: "VBA Mode" where the default becomes Private? (just thinking aloud here)

**Reply by KDGundermann** [reply]
@mburns08109 I hope so, that we will be able to use TB Librarys (ActiveX?) in VBA (Microsoft Access)
I have huge amount of library functions in *.accdb files and I hope I can port them to TB (for better speed)
